### PR TITLE
Add array arithmetic

### PR DESCRIPTION
**Pull request by tpoterba**
Enable +, -, /, * operations for:
 - numeric arrays and numeric arrays
 - scalars and numeric arrays
 - numeric arrays and scalars

### CODE REVIEW

**Comment by tpoterba** [comment]
okay, now ready.  In addition to the bugfix and expr array features, fixed a lingering stupid problem in our error traversal method

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala [hunk 1]
```diff
@@ -213,6 +213,37 @@ object AST extends Positional {
         null
     }
   }
+
+  def evalComposeNumericArray[T](ec: EvalContext, subexpr: AST)(g: (IndexedSeq[T]) => Any)
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: What about add versions of evalCompose for a scalar and an array?

Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: yeah, this is slightly better.  But I'd really like to mix and match arrays and scalars, and even if there's an evalCompose for arrays, that doesn't solve the problem.

File: src/main/scala/org/broadinstitute/hail/driver/Main.scala [hunk 2]
```diff
@@ -95,10 +95,11 @@ object Main {
   }
 
 
-  def expandException(cmd: Command, e: Throwable): String =
-    s"${ e.getClass.getName }: ${ e.getMessage }\n\tat ${ e.getStackTrace.mkString("\n\tat ") }${
-      Option(e.getCause).foreach(exception => expandException(cmd, exception))
+  def expandException(cmd: Command, e: Throwable): String = {
+    s"${ e.getClass.getName }: ${ e.getLocalizedMessage }\n\tat ${ e.getStackTrace.mkString("\n\tat ") }${
+      Option(e.getCause).map(exception => expandException(cmd, exception)).getOrElse("")
```
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/driver/Main.scala]: I want to draw attention to this and how silly this mistake was

Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/driver/Main.scala]: also I take responsibility for that

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala (hunk 1 above)
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: I'll think on this a bit and try to come up with a proposal.

Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: This is a good change, let's get it in.  I also think the expr code has gotten completely out of hand.  I was debating if it is even worth adding new evalComposes.  Let's have a meeting with @danking this week and start to flesh out a plan towards HailQL + codegen.

Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: actually, I think this doesn't handle missing properly.  Give me a little time to think on it

Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]:
```
        case (t: TNumeric, TArray(_), TDouble) => AST.evalCompose[Any, IndexedSeq[_]](ec, lhs, rhs) { case (num, arr) =>
          val d = DoubleNumericConversion.to(num)
          arr.map(x => d - DoubleNumericConversion.to(x))
        }
```
Missing value in the array will cause a match error

Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: Indeed.  The whole point of evalCompose is to handle missing.

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala [hunk 3]
```diff
@@ -1354,7 +1385,7 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
         m.mapValues { elt =>
           localA(localIdx) = elt
           bodyFn()
-        }
+        }.force
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: This should be a separate PR if possible.

File: src/test/scala/org/broadinstitute/hail/methods/ExprSuite.scala [hunk 4]
```diff
@@ -383,6 +383,15 @@ class ExprSuite extends SparkSuite {
     assert(D_==(eval[Double]("sqrt(5.6)").get, math.sqrt(5.6)))
     assert(D_==(eval[Double]("pow(2, 3)").get, 8.0))
 
+    assert(eval[IndexedSeq[Int]]("""[1,2,3] + [2,3,4] """).contains(IndexedSeq(3, 5, 7)))
+    assert(eval[IndexedSeq[Int]]("""[1,2,3] - [2,3,4] """).contains(IndexedSeq(-1, -1, -1)))
+    assert(eval[IndexedSeq[Double]]("""[1,2,3] / [2,3,4] """).contains(IndexedSeq(.5, 2.toDouble / 3.toDouble, .75)))
+    assert(eval[IndexedSeq[Int]]("""[1] + [2,3,4] """).contains(IndexedSeq(3)))
```
Comment (danking) [inline src/test/scala/org/broadinstitute/hail/methods/ExprSuite.scala]: This behavior makes me uneasy because it masks mistakes.

File: src/test/scala/org/broadinstitute/hail/methods/ExprSuite.scala [hunk 5]
```diff
@@ -383,6 +383,15 @@ class ExprSuite extends SparkSuite {
     assert(D_==(eval[Double]("sqrt(5.6)").get, math.sqrt(5.6)))
     assert(D_==(eval[Double]("pow(2, 3)").get, 8.0))
 
+    assert(eval[IndexedSeq[Int]]("""[1,2,3] + [2,3,4] """).contains(IndexedSeq(3, 5, 7)))
```
Comment (danking) [inline src/test/scala/org/broadinstitute/hail/methods/ExprSuite.scala]: Instead of `contains` should be we using `equals`?

Comment (tpoterba) [inline src/test/scala/org/broadinstitute/hail/methods/ExprSuite.scala]: Contains is called on the Option[T] result of evaluate.

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala [hunk 6]
```diff
@@ -1631,6 +1631,191 @@ case class BinaryOp(posn: Position, lhs: AST, operation: String, rhs: AST) exten
       s.r.findFirstIn(t).isDefined
     }
 
+    case ("+", TArray(elementType)) =>
+      ((lhs.`type`, rhs.`type`, elementType): @unchecked) match {
+        case (t: TNumeric, TArray(_), TDouble) => AST.evalCompose[Any, IndexedSeq[_]](ec, lhs, rhs) { case (num, arr) =>
+          val d = DoubleNumericConversion.to(num)
+          arr.map(x => DoubleNumericConversion.to(x) + d)
+        }
```
Comment (danking) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: I suspect we can abstract a bit more here using these functions:

```scala
  def liftArrayOpToEval[T](ec: EvalContext, lhs: AST, rhs: AST, coerce: Any => T, binOp: (T, T) => T): () => Any =
    liftEval[Any, IndexedSeq[Any]](ec, lhs, rhs, liftArrayOp(coerce, binOp))

  def liftArrayOp[T](coerce: Any => T, binOp: (T, T) => T): (Any, IndexedSeq[Any]) => IndexedSeq[T] =
    { case (num, arr) => arr.map(x => binOp(coerce(x), coerce(num))) }

  def liftEval[T1, T2](ec: EvalContext, lhs: AST, rhs: AST, g: (T1, T2) => Any): () => Any =
    AST.evalCompose[Any, IndexedSeq[_]](ec, lhs, rhs)(g)
```

which should permit defining this case as:

```scala
case (t: TNumeric, TArray(_), TDouble) => liftArrayOpToEval[Double](ec, lhs, rhs, DoubleNumericConversion.to, { _ + _})
```

Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: This actually doesn't handle missing, unless `[T]` is explicitly `Option[S]`

Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: Would get NullPointerException here if an array element is null: 
```
{ case (num, arr) => arr.map(x => binOp(coerce(x), coerce(num))) }
```

Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: this also coerces the scalar once per array element, which is inefficient.  I think we should have a framework where the efficiency and abstraction are both easily specified, but that is definitely not an easy thing.

Comment (danking) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: re: the above comments and some offline discussions, here's an update `liftArrayOp`:
```scala
  def liftArrayOp[T >: Null](coerce: Any => T, binOp: (T, T) => T): (Any, IndexedSeq[Any]) => IndexedSeq[T] =
    { case (num, arr) => {
      val y = coerce(num);
      arr.map(x => if (x == null) null else binOp(coerce(x), y))
    } }
```

**Comment by tpoterba** [comment]
Greatly simplified the evaluation matching.

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala [hunk 7]
```diff
@@ -1631,6 +1632,65 @@ case class BinaryOp(posn: Position, lhs: AST, operation: String, rhs: AST) exten
       s.r.findFirstIn(t).isDefined
     }
 
+    case ("+" | "*" | "-" | "/", TArray(elementType)) =>
+      val f: (Any, Any) => Any = (operation, elementType) match {
```
Comment (danking) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: I sent you a PR that simplifies this a tad more

https://github.com/tpoterba/hail/pull/5

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala [hunk 8]
```diff
@@ -1622,6 +1623,12 @@ case class Let(posn: Position, bindings: Array[(String, AST)], body: AST) extend
 }
 
 case class BinaryOp(posn: Position, lhs: AST, operation: String, rhs: AST) extends AST(posn, lhs, rhs) {
+  def liftBinaryWithCoerce[T,U,V](coerce: T => U)(op: (U, U) => V): (T, T) => V =
+    (l, r) => op(coerce(l), coerce(r))
+
+  def threadNulls(f: (Any, Any) => Any): (Any, Any) => Any =
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: These are basically trying to implement `evalComposeNumeric` without the compose.  I propose:

```
  def evalNumeric[T1, T2](g: (T1, T2) => Any): (Any, Any) => Any
```

with any necessary implicits.  Then you can do `evalNumeric[Double, Double](_ + _)`  analogously to the scalar implementation of +, etc.

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala [hunk 9]
```diff
@@ -1631,6 +1638,40 @@ case class BinaryOp(posn: Position, lhs: AST, operation: String, rhs: AST) exten
...
+        case ("-", TFloat) => liftBinaryWithCoerce(FloatNumericConversion.to)({ _ - _ })
+        case ("/", TDouble) => liftBinaryWithCoerce(DoubleNumericConversion.to)({ _ / _ })
+      }
+      val f = threadNulls(nullsUnhandled)
+
+      ((lhs.`type`, rhs.`type`): @unchecked) match {
+        case (t: TNumeric, TArray(_)) => AST.evalCompose[Any, IndexedSeq[_]](ec, lhs, rhs) { case (num, arr) =>
+          arr.map(elem => f(num, elem))
+        }
+        case (TArray(_), t: TNumeric) => AST.evalCompose[IndexedSeq[_], Any](ec, lhs, rhs) { case (arr, num) =>
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: I'd put the 2-array case first, then `t: TNumeric` can just `_`.

**Comment by cseed** [comment]
Few minor comments, looks good.  Needs rebase.  Also, I'm going to be a ****: break this into three PRs: the `.force` fix, the exception fix, and the array changes.

File: src/main/scala/org/broadinstitute/hail/expr/AST.scala (hunk 8 above)
Comment (danking) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: I was thinking:

```scala
  def evalNumeric[T1, T2](g: (T1, T2) => Any): (() => Any, () => Any) => Any
```

which could be used to implement `evalComposeNumeric`. We could also expose a non-thunked version.

Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: Yeah, that's nice.  Tim needs the non-thunked version because he's applying this to zipped arrays.

Comment (danking) [inline src/main/scala/org/broadinstitute/hail/expr/AST.scala]: There's a monad here somewhere.
